File: MyMovies-MVP/app/src/main/java/com/jbrunton/mymovies/search/SearchResultsAdapter.java
package com.jbrunton.mymovies.search;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.jbrunton.mymovies.BaseRecyclerAdapter;
import com.jbrunton.mymovies.R;
import com.squareup.picasso.Picasso;

public class SearchResultsAdapter extends BaseRecyclerAdapter<SearchItemViewState, SearchResultsAdapter.ViewHolder> {
    public SearchResultsAdapter(Context context, int layoutId) {
        super(layoutId, new ViewHolderFactory(context));
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        private TextView titleView;
        private TextView releaseDateView;
        private ImageView poster;
        private TextView ratingView;

        ViewHolder(View itemView) {
            super(itemView);
            titleView = (TextView) itemView.findViewById(R.id.title);
            releaseDateView = (TextView) itemView.findViewById(R.id.release_date);
            ratingView = (TextView) itemView.findViewById(R.id.rating);
            poster = (ImageView) itemView.findViewById(R.id.poster);
        }
    }

    protected static class ViewHolderFactory implements BaseRecyclerAdapter.ViewHolderFactory<SearchItemViewState, ViewHolder> {
        private final Context context;

        public ViewHolderFactory(Context context) {
            this.context = context;
        }

        @Override
        public ViewHolder createViewHolder(View view) {
            return new ViewHolder(view);
        }

        @Override
        public void bindHolder(ViewHolder holder, SearchItemViewState item) {
            holder.titleView.setText(item.title());
            if (holder.releaseDateView != null) {
                holder.releaseDateView.setText(item.yearReleased());
            }
            if (holder.ratingView != null) {
                holder.ratingView.setText(Html.fromHtml(item.rating()));
            }
            Picasso.with(context)
                    .load(item.posterUrl())
                    .resize(185, 275)
                    .centerCrop()
                    .into(holder.poster);
        }
    }
}

File: app/src/main/java/com/jbrunton/mymovies/ui/search/SearchResultsAdapter.kt
package com.jbrunton.mymovies.ui.search

import android.content.Context
import android.content.Intent
import android.text.Html
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.jbrunton.mymovies.R
import com.jbrunton.mymovies.helpers.PicassoHelper
import com.jbrunton.mymovies.ui.moviedetails.MovieDetailsActivity
import com.jbrunton.mymovies.ui.movies.MovieSearchResultViewState
import com.jbrunton.mymovies.ui.shared.BaseRecyclerAdapter

class SearchResultsAdapter(
        context: Context,
        layoutId: Int
) : BaseRecyclerAdapter<MovieSearchResultViewState, SearchResultsAdapter.ViewHolder>(
        layoutId,
        ViewHolderFactory(context)
) {

    class ViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {
        val titleView: TextView
        val releaseDateView: TextView?
        val poster: ImageView
        val ratingView: TextView?
        val movieCardView: View
        val divider: View?

        init {
            movieCardView = itemView.findViewById(R.id.movie_card_view)
            titleView = itemView.findViewById(R.id.title)
            releaseDateView = itemView.findViewById(R.id.release_date)
            ratingView = itemView.findViewById(R.id.rating)
            poster = itemView.findViewById(R.id.poster)
            divider = itemView.findViewById<View?>(R.id.divider)
        }
    }

    class ViewHolderFactory(private val context: Context) : BaseRecyclerAdapter.ViewHolderFactory<MovieSearchResultViewState, ViewHolder> {

        private val picassoHelper = PicassoHelper()

        override fun createViewHolder(view: View): ViewHolder {
            return ViewHolder(view)
        }

        override fun bindHolder(holder: ViewHolder, item: MovieSearchResultViewState, items: List<MovieSearchResultViewState>, position: Int) {
            holder.titleView.text = item.title
            if (holder.releaseDateView != null) {
                holder.releaseDateView.text = item.yearReleased
            }
            if (holder.ratingView != null) {
                holder.ratingView.text = Html.fromHtml(item.rating)
            }
            picassoHelper.loadSearchResultImage(context, holder.poster, item.posterUrl)
            holder.movieCardView.setOnClickListener {
                val intent = Intent(context, MovieDetailsActivity::class.java)
                intent.putExtra("MOVIE_ID", item.movieId)
                context.startActivity(intent)
            }
            if (holder.divider != null) {
                holder.divider.visibility = if (position < items.size - 1) {
                    View.VISIBLE
                } else {
                    View.GONE
                }
            }
        }
    }
}
